git-index/src/access.rs

```diff
@@ -1,6 +1,83 @@
+use crate::{
+    entry::{self, Flags, Mode, Stat, Time},
+    extension, Entry, PathStorage, State, Version,
+};
 use bstr::{BStr, ByteSlice};
+use git_object::{tree::EntryMode, TreeRefIter};
+use git_traverse::tree::breadthfirst;
 
-use crate::{entry, extension, Entry, PathStorage, State, Version};
+/// initialization
+impl State {
+    /// Takes in an oid of a tree object and creates and returns a [`State`][git_index::State] from its children.
+    pub fn from_tree<Find>(tree: &git_hash::oid, mut find: Find) -> Result<Self, breadthfirst::Error>
+    where
+        Find: for<'a> FnMut(&git_hash::oid, &'a mut Vec<u8>) -> Option<TreeRefIter<'a>>,
+    {
+        let mut buf = Vec::new();
+        let root = find(tree, &mut buf).expect("couldn't find a tree for given oid");
+        let state = breadthfirst::State::default();
+        let mut delegate = git_traverse::tree::Recorder::default();
+        breadthfirst(root, state, &mut find, &mut delegate)?;
+
+        // TODO: estimate size
+        let mut path_backing: PathStorage = Vec::new();
+
+        // TODO: insert entries directly into the correct position
```

Comment:
If you did a depth-first traversal you'd naturally end up with entries in index order, wouldn't you? Are you doing depth first for any particular reason? (I'm not familiar with how other implementations do this — please excuse my ignorance if this is something well known. 😅)

Reply:
I mainly used breadth-first traversal as this was the only algorithm available in the git-traverse crate at the moment 😄
But I believe you are right that depth-first would yield entries in the correct order directly, provided the TreeObject's data is already sorted.

Reply:
I think I remember now why breadth-first was implemented. It was all about having only two buffers to store object data. Depth first should be no problem using a free-list of buffers for the objects to (re-)use. I still wonder if this 'special sorting' that entry-paths are using is going to spoil the fun, but it's most definitely worth finding out.

Reply:
I just remembered a gotcha that I ran into with libgit2 that may inform what you do here:

Even though the Git docs say the index is in memcmp order, in practice I have found that it differs by operating system / filesystem. On macOS with default configuration, for example, I found that the index I got out of libgit2 was ordered with a **case-insensitive** sort, whereas on Linux it was case-sensitive. This resulted in a really sneaky bug in a program I was writing that only showed up on macOS.

I'm afraid I can't remember whether this was just an issue with libgit2 or if the on-disk index created by Git itself was also not in the documented order. Either way, it seems worth being very careful here.

Reply:
Thanks for sharing! This sounds like the case-sensitivity configuration of the repository is influencing this. Ultimately we typically look at the `git` source code to gain clarity, and definitely will do here to not potentially replicate `libgit2` bugs 😅.